### PR TITLE
Add CI script to deploy published releases

### LINKED ISSUE #1936: Add CI script to deploy published releases
We need a script that will be used in periodics to test our published images. It essentially codifies what needs to be done to deploy a particular scylla-operator version.

### PR DESCRIPTION
**Pull request by tnozicka**
**Description of your changes:**
This script will be used in periodics to test our published images. It essentially codifies what needs to be done to deploy a particular scylla-operator version. Some action could be simplified by restructuring manifest but it's out of the scope of this PR. Eventually, we should fix this and possibly have some smarter templating / publish manifests into releases with the replaced image. So take this PR as a start with what we have now, not where it should end.

**Which issue is resolved by this Pull Request:**
Resolves #1936 

### Requires
- [x] #1901 
- [x] #1922
- [x] https://github.com/scylladb/scylla-operator-release/pull/193
- [x] https://github.com/scylladb/scylla-operator-release/pull/194

### CODE REVIEW

**Comment by tnozicka** [comment]
this can't be used by presubmits by definition as their merged git shas are not published - this is only for published releases

**Comment by rzetelskik** [comment]
> this can't be used by presubmits by definition as their merged git shas are not published - this is only for published releases

I know, but the question was about how is this going to be integrated with CI, pointing at whether this PR is potentially missing any adjustments.

**Comment by tnozicka** [comment]
> I know, but the question was about how is this going to be integrated with CI, pointing at whether this PR is potentially missing any adjustments.

This will be used only by new periodics based on already published images (we don't have these periodics yet). 
(Eventually, we can add an on demand presubmit using this script against a latest release to smoke test it.)

**Comment by tnozicka** [comment]
added the extra presubmit in https://github.com/scylladb/scylla-operator-release/pull/157

**Review by zimnx — APPROVED** [review]
/lgtm

please fix CI script
/hold

**Comment by rzetelskik** [comment]
https://github.com/scylladb/scylla-operator-release/pull/163

**Comment by rzetelskik** [comment]
`time="2024-03-28T13:38:18Z" level=fatal msg="Error parsing image name \"docker://docker.io/scylladb/scylla-operator:latest@sha256:0c8896a666fdc0d273b92dbd1bd936140d1da34e6c25b9f7ac1260a7acc18f16\": Docker references with both a tag and digest are currently not supported"`

needs a fix

**Comment by tnozicka** [comment]
fixed in https://github.com/scylladb/scylla-operator-release/pull/168

**Comment by tnozicka** [comment]
https://github.com/scylladb/scylla-operator-release/pull/168 has landed
/retest

**Comment by tnozicka** [comment]
cluster didn't spin up in time
/retest

**Comment by tnozicka** [comment]
oh, I've tested this but seems like it depends on `extglob` which I had enabled and the image doesn't. need to fix that in the CI config

**Comment by tnozicka** [comment]
https://github.com/scylladb/scylla-operator-release/pull/180 landed
/retest

**Comment by tnozicka** [comment]
(this need more work on artifact collection and specifying nodeconfig on CI side)

File: hack/ci-deploy.sh
Comment (rzetelskik) [inline hack/ci-deploy.sh:18]: I assume this is coming from https://github.com/scylladb/scylla-operator/pull/1922, please rebase

File: hack/ci-deploy-release.sh
Comment (rzetelskik) [inline hack/ci-deploy-release.sh:66]: is this missing the tweaks from ci-deploy (https://github.com/scylladb/scylla-operator/blob/master/hack/ci-deploy.sh#L34-L35) on purpose? If so, is it superfluous there or accidentally missing here?

File: hack/ci-deploy-release.sh
Comment (rzetelskik) [inline hack/ci-deploy-release.sh:43]: missing wait-for-object-creation from ci-deploy: https://github.com/scylladb/scylla-operator/blob/b7ca9a4fce4aa060cec0bb8139382fdf8d909928/hack/ci-deploy.sh#L48. Is it superfluous there or accidentally missing here?

File: hack/ci-deploy-release.sh
Comment (rzetelskik) [inline hack/ci-deploy-release.sh:77]: diverges from ci-deploy: https://github.com/scylladb/scylla-operator/blob/b7ca9a4fce4aa060cec0bb8139382fdf8d909928/hack/ci-deploy.sh#L60-L64

File: hack/ci-deploy-release.sh
Comment (rzetelskik) [inline hack/ci-deploy-release.sh:107]: ditto https://github.com/scylladb/scylla-operator/blob/b7ca9a4fce4aa060cec0bb8139382fdf8d909928/hack/ci-deploy.sh#L81-L84

**Review by rzetelskik — CHANGES_REQUESTED** [review]

File: hack/ci-deploy.sh
Comment (tnozicka) [reply hack/ci-deploy.sh]: weird, fixed with a clean rebase

File: hack/ci-deploy-release.sh
Comment (tnozicka) [reply hack/ci-deploy-release.sh:99]: I don't think those have to match as these are different use cases. This should be eventually dropped and all our manifest should pick the correct storage class without relying on the defaults or needing messing with the global classes. But our provisioner doesn't support the full disks yet, so we are not there yet.

File: hack/ci-deploy-release.sh
Comment (tnozicka) [reply hack/ci-deploy-release.sh]: done

File: hack/ci-deploy-release.sh
Comment (tnozicka) [reply hack/ci-deploy-release.sh:109]: nodeconfigs are waited for a bit earlier
`find` doesn't work with `https://`

**Comment by tnozicka** [comment]
@rzetelskik please have a look at the comment replies, I'd like to land this and unblock your arm64 PR

File: hack/ci-deploy-release.sh
Comment (rzetelskik) [reply hack/ci-deploy-release.sh:80]: it shouldn't be defined here then? https://github.com/scylladb/scylla-operator/pull/1837/files#diff-5da3c104525c31203c7505086de76d48b279733088c4ec836270ee1aabcde637R23

File: hack/ci-deploy-release.sh
Comment (rzetelskik) [reply hack/ci-deploy-release.sh:99]: Imo having them diverge jut makes it harder to find the root cause if anything fails, and doing the same thing for the regular ci-deploy would be preferred?

File: hack/ci-deploy-release.sh
Comment (rzetelskik) [reply hack/ci-deploy-release.sh:43]: should it be removed from the other script then? (not in this pr)

**Comment by rzetelskik** [comment]
> @rzetelskik please have a look at the comment replies, I'd like to land this and unblock your arm64 PR

done

File: hack/ci-deploy-release.sh
Comment (tnozicka) [reply hack/ci-deploy-release.sh:80]: good catch, that one shouldn't be there

Comment (tnozicka) [reply hack/ci-deploy-release.sh:80]: fixed

File: hack/ci-deploy-release.sh
Comment (tnozicka) [reply hack/ci-deploy-release.sh:99]: only temporarily, I've added an explicit mention to clean this up in https://github.com/scylladb/scylla-operator/issues/1554

File: hack/ci-deploy-release.sh
Comment (tnozicka) [reply hack/ci-deploy-release.sh:43]: I think so

Comment (rzetelskik) [reply hack/ci-deploy-release.sh:43]: should we have a tracker for this? Even if it's in the backlog

**Review by rzetelskik — APPROVED** [review]
lgtm
/assign zimnx
in case you want to have a second look at it, it changed since your review. Otherwise feel free to remove the hold.

File: hack/ci-deploy-release.sh
Comment (tnozicka) [reply hack/ci-deploy-release.sh:43]: https://github.com/scylladb/scylla-operator/pull/1945

File: hack/ci-deploy-release.sh
Comment (zimnx) [inline hack/ci-deploy-release.sh:102]: nit: these could be replaced with waiting for our three aggregated conditions on ScyllaCluster without hardcoding names of Operator managed resources.

**Review by zimnx — APPROVED** [review]
One nit. Feel free to cancel hold even without applying, i'm fine with having both options. 

/lgtm
/hold

File: hack/ci-deploy-release.sh
Comment (tnozicka) [reply hack/ci-deploy-release.sh:102]: It's a good point but it's in ci-deploy.sh as well, I'll send a followup.
/hold cancel

Comment (tnozicka) [reply hack/ci-deploy-release.sh:102]: https://github.com/scylladb/scylla-operator/pull/1947

**Pull request merged**
